### PR TITLE
feat: SBOM linking

### LINKED ISSUE #81: SBOM "link" feature
## Summary
Given one or more SBOM documents, create a third SBOM document containing external references to those documents.
Something like:
```
bomctl link sbom1.json sbom2.json ... -o linked_sbom.json
```
Ideally, bomctl should also be able to express relationships between the components in the sbom:
```
bomctl link relationship=from:sbom1.json,to:sbom2.json,type:contains
```
## Use Case
Suppose a base container image has an SBOM
```mermaid
graph LR
    A(Distroless Creators)
    B(Distroless container image)
    C(Distroless SBOM)
    A --> B
    A --> C
```
And a Go binary has its own SBOM
```mermaid
graph LR
    A(Go Binary Creators)
    B(Go Binary)
    C(Go binary SBOM)
    A --> B
    A --> C
```
During containerization of the Go binary, I would also like to create an SBOM for this container which has external references to the base container's SBOM and the go binary's SBOM
```mermaid
graph LR
    A(Distroless Container)
    B(Go Binary)
    C(Containerized Go Binary)
    A -->|FROM distroless:latest| C
    B --> |COPY go_binary|C

    D(Containerized Go Binary SBOM)
    E(Distroless Container SBOM)
    F(Go Binary SBOM)

    D --> |contains|E
    D --> |contains|F
    F --> |dependsOn|E
```

### PR DESCRIPTION
**Pull request by jhoward-lm**
## Description

This PR implements the `link` subcommand.

### Miscellaneous

- VSCode snippet to generate `testify` suite

Closes #81

## Type of change

- [X] New feature (non-breaking change which adds functionality)

## Checklist

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [X] Any dependent changes have been merged and published in downstream modules
- [X] I have checked my code and corrected any misspellings

### CODE REVIEW

**Review by ashearin — COMMENTED** [review]
Still need to mess with it a bit, only thing I'm not seeing that I'd expect is updates to documentation

**Comment by ashearin** [comment]
I guess there something else also, are we adding e2e tests for this cmd with this PR or create a follow on issue to do that?

**Comment by ashearin** [comment]
Outside of the scope of this task, but would it make sense to have a flag for `bomctl list <sbom-id>` to list the nodes of that document, maybe just the name and native id? Potentially have to add a filter flag as well since some of these documents are quite long. Just an idea.  (Speaking about base list cmd specifically, not `bomctl link list`)

File: internal/pkg/link/link.go [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+	links := tree.Root(fmt.Sprintf("Links for %s %s:", fromType, style.Render(id))).
+		Child(children).
+		EnumeratorStyle(style)
+
+	fmt.Fprintln(os.Stdout, links)
+
+	return nil
+}
+
+func RemoveLink(backend *db.Backend, opts *options.LinkOptions) error {
```
Comment (ashearin) [inline internal/pkg/link/link.go:122]: Having some trouble getting this working: 

`link add` and `link list` worked fine, was able to add a link, then view it, but I can't seem to remove it: 
``` bash
$ go run . link remove document:urn:uuid:f360ad8b-dc41-4256-afed-337a04dff5db document:https://anchore.com/syft/file/bomctl_0.3.0
_linux_amd64.tar.gz-1b838d44-9d3c-47d0-9f7f-846397e701fa#DOCUMENT

$ go run . link list document:urn:uuid:f360ad8b-dc41-4256-afed-337a04dff5db
Links for document urn:uuid:f360ad8b-dc41-4256-afed-337a04dff5db:
└──https://anchore.com/syft/file/bomctl_0.3.0_linux_amd64.tar.gz-1b838d44-9d3c-47d0-9f7f-846397e701fa#DOCUMENT

$ go run . link remove document:https://anchore.com/syft/file/bomctl_0.3.0_linux_amd64.tar.gz-1b838d44-9d3c-47d0-9f7f-846397e701f
a#DOCUMENT document:urn:uuid:f360ad8b-dc41-4256-afed-337a04dff5db

$ go run . link list document:urn:uuid:f360ad8b-dc41-4256-afed-337a04dff5db
Links for document urn:uuid:f360ad8b-dc41-4256-afed-337a04dff5db:
└──https://anchore.com/syft/file/bomctl_0.3.0_linux_amd64.tar.gz-1b838d44-9d3c-47d0-9f7f-846397e701fa#DOCUMENT
```

As you can see I tried swapping the document ids, but still nothing. Might be missing something.

File: internal/pkg/link/link.go [hunk 2]
```diff
@@ -0,0 +1,151 @@
+// -----------------------------------------------------------------------------
```
Comment (ashearin) [inline internal/pkg/link/link.go:1]: We could probably stand to add some more logging to this file to give insight on what is happening.

File: internal/pkg/link/link.go (hunk 1 above)
Comment (ashearin) [reply internal/pkg/link/link.go:122]: Clear seems to work fine:

```bash 
$ go run . link clear document:urn:uuid:f360ad8b-dc41-4256-afed-337a04dff5db
$ go run . link list document:urn:uuid:f360ad8b-dc41-4256-afed-337a04dff5db
WARN  link: No links found for document id=urn:uuid:f360ad8b-dc41-4256-afed-337a04dff5db
```

Comment (jhoward-lm) [reply internal/pkg/link/link.go:112]: Good catch! Just pushed a fix, give that a shot

Comment (ashearin) [reply internal/pkg/link/link.go:112]: Works! Thanks!

**Comment by jhoward-lm** [comment]
> I guess there is something else, are we adding e2e tests for this cmd with this PR or create a follow on issue to do that?

I would imagine as part of this one

**Review by lallevato-lm — CHANGES_REQUESTED** [review]

File: internal/pkg/link/link.go [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+	linkIDs := sliceutil.Extract(annotations, func(annotation *ent.Annotation) string {
+		return annotation.Value
+	})
+
+	writeLinksTree(id, fromType, linkIDs, documents, backend)
+
+	return nil
+}
+
+func RemoveLink(backend *db.Backend, opts *options.LinkOptions) error {
```
[the remark this answers is not in the document]
Comment (jhoward-lm) [reply internal/pkg/link/link.go:154]: @lallevato-lm Added some logger calls to state what was removed. Does that resolve this one?

File: internal/pkg/link/link.go [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+
+		if err := backend.AddNodeAnnotations(opts.NodeIDs[0], db.LinkToAnnotation, toID); err != nil {
+			return fmt.Errorf("%w", err)
+		}
+	}
+
+	return nil
+}
+
+func ClearLinks(backend *db.Backend, opts *options.LinkOptions) error {
```
[the remark this answers is not in the document]
Comment (jhoward-lm) [reply internal/pkg/link/link.go:69]: @lallevato-lm Added some logger calls to state what was removed from each document/node as part of the clear operation. Does that resolve this one?

**Review by lallevato-lm — APPROVED** [review]

**Review by lallevato-lm — APPROVED** [review]

File: internal/e2e/link/link.txtar
Comment (lallevato-lm) [inline internal/e2e/link/link.txtar:1]: Do we want a use case where bomctl tries to do a link operation on a document which doesn't exist?

Comment (lallevato-lm) [reply internal/e2e/link/link.txtar:1]: Looks good!

**Review by lmphil — APPROVED** [review]

**Comment by lmphil** [comment]
Looks good. Only missing changes to the README for the new command.

**Comment by jhoward-lm** [comment]
> Looks good. Only missing changes to the README for the new command.

@lmphil README updated

**Pull request merged**